SQL: Fix wrong results when sorting on aggregate

matriv opened the pull request:
- Previously, when shorting on an aggregate function the bucket
processing ended early when the explicit (LIMIT XXX) or the impliciti
limit of 512 was reached. As a consequence, only a set of grouping
buckets was processed and the results returned didn't reflect the global
ordering.

- Previously, the priority queue shorting method had an inverse
comparison check and the final response from the priority queue was also
returned in the inversed order because of the calls to the `pop()`
method.

Fixes: #42851

matriv committed:
SQL: Fix wrong results when sorting on aggregate

- Previously, when shorting on an aggregate function the bucket
processing ended early when the explicit (LIMIT XXX) or the impliciti
limit of 512 was reached. As a consequence, only a set of grouping
buckets was processed and the results returned didn't reflect the global
ordering.

- Previously, the priority queue shorting method had an inverse
comparison check and the final response from the priority queue was also
returned in the inversed order because of the calls to the `pop()`
method.

Fixes: #42851

elasticmachine commented:
Pinging @elastic/es-search

jimczi commented on x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/cli/ErrorsTestCase.java:
```diff
@@ -97,8 +97,15 @@ public void testSelectScoreInScalar() throws Exception {
+    public void testHardLimitForSortOnAggregate() throws Exception {
+        index("test", body -> body.field("a", 1).field("b", 2));
+        String commandResult = command("SELECT max(a) max FROM test GROUP BY b ORDER BY max LIMIT 10000");
+        assertEquals(START + "Bad request [[3;33;22mThe maximum LIMIT for aggregate sorting is [512], received [10000]" + END,
```
Unrelated to this pr but I don't think we should treat ordered GROUP_BY differently than natural GROUP_BY (sorted by keys). We should use the same hard limit for both, the cost of an ordered GROUP_BY depends greatly on the cardinality of the groups and not so much on the size of the priority queue (assuming it remains reasonable ;)).

matriv replied:
I don't think I agree here. When we don't have an order by on aggregate then we don't need to keep this extra buffer in memory (PriorityQueue) in the SQL module, instead we have a cursor from ES to paginate the results, thus we don't need to apply a hard limit. Or maybe I'm missing something?

jimczi replied:
The memory to keep the best buckets in the PriorityQueue should be identical to the one that is used in ES when we compute the top hits so it's only a question on where this memory is held. I understand that the priority queue in the client can increase the memory but it shouldn't make a big difference if the size is 512 or 1000 or even 10,000. Most of the time will be spent on paginating the buckets and a limit of 1 or 512 shouldn't make any difference.

matriv replied:
Thx, opened an issue to track it: https://github.com/elastic/elasticsearch/issues/43168

jimczi replied:
My thinking here is the other way around, we should increase the hard limit of the ordered group_by to match the one that we set to "normal" group_by. `10,000` seems a reasonable value IMO.

matriv replied:
+1, I misunderstood you here. Changed the issue. Thx!

jimczi commented on x-pack/plugin/sql/qa/src/main/java/org/elasticsearch/xpack/sql/qa/cli/FetchSizeTestCase.java:
```diff
@@ -49,4 +49,32 @@ public void testInvalidFetchSize() throws IOException {
+        Request request = new Request("PUT", "/test/_bulk");
+        request.addParameter("refresh", "true");
+        StringBuilder bulk = new StringBuilder();
+        for (int i = 1; i <= 5000; i++) {
```
This test might be slow, right ? I think you can check that a GROUP_BY on groups with a cardinality greater than the limit and the size works with less than 5000 documents. Is there a way to set the internal size that is used to paginate over the buckets in the sorted group_by ?

matriv replied:
It's not that slow and I used a size of 5000 to "escape" from any bugs regarding setting the fetch (defaults to 1000) that could hide this bug. If you still think it's too much I can change to something like 100, limit 20, fetch size 4.

jimczi replied:
+1, this is why I asked if you can modify the fetch size here. In such case `100` seems enough

jimczi commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java:
```diff
@@ -248,31 +205,22 @@ private void doResponse(RowSet rowSet) {
 
-        private boolean consumeRowSet(RowSet rowSet) {
+        private void consumeRowSet(RowSet rowSet) {
             // use a synchronized block for visibility purposes (there's no concurrency)
```
remove the comment now that the synchronized block is gone ?

jimczi reviewed (commented):
Good catch @matriv ;) , I left some minor comments but the fix looks good to me.

matriv committed:
Address comments

matriv commented:
Thx @jimczi, addressed comments.

jimczi reviewed (approved):
LGTM, thanks @matriv

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java:
```diff
@@ -248,31 +205,21 @@ private void doResponse(RowSet rowSet) {
-            // use a synchronized block for visibility purposes (there's no concurrency)
+        private void consumeRowSet(RowSet rowSet) {
             ResultRowSet<?> rrs = (ResultRowSet<?>) rowSet;
-            synchronized (data) {
```
Why removing the `synchronized`?

matriv replied:
There's no concurrency here, synchronized will add locking overhead without need.

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/execution/search/QuerierTests.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @SuppressWarnings("rawtypes")
+    public void testAggSorting_TwoFields() {
```
Would be difficult to add a more randomized set of tests? For example, 3 comparators with random ASC/DESC ordering and a more-or-less random short set of int values (random between 0 and 1 for all three values).

For example:
1,2,3,4,5,6,7
0,0,1,0,1,1,0 ASC
0,0,0,0,1,1,0 ASC
0,1,1,0,0,1,1 DESC

Would yield:
2 (0,0,1)
7 (0,0,1)
1 (0,0,0)
4 (0,0,0)
3 (1,0,1)
6 (1,1,1)
5 (1,1,0)

matriv replied:
Thx for the suggestion! Added the test which provides much more confidence regarding implementation correctness.

astefan reviewed (commented):
Great work!
Left some comments.
Also, are there any tests that combine both ES ordering and client-side ordering?

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/CompositeAggsRowSet.java:
```diff
@@ -40,18 +40,22 @@
 
         // page size
-        size = limit < 0 ? buckets.size() : Math.min(buckets.size(), limit);
+        size = limit == -1 ? buckets.size() : Math.min(buckets.size(), limit);
```
It will also work, but we introduced the `-1` which means "no limit - need to process all buckets", and it's the only negative value we can get now. If you check [here](https://github.com/elastic/elasticsearch/pull/43154/files/8ea19efa6df6ca80b02d31d61f05b598d7b4d854#diff-c4c82fd30c72e84bb90b408910d43a4eR52) you'll see that when we want to stop we set remaining limit to `0`. So to sum up, I changed the check to `== -1` for more clarity.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java:
```diff
@@ -248,31 +205,21 @@ private void doResponse(RowSet rowSet) {
+            for (boolean hasRows = rrs.hasCurrentRow(); hasRows; hasRows = rrs.advanceRow()) {
+                List<Object> row = new ArrayList<>(rrs.columnCount());
+                rrs.forEachResultColumn(row::add);
+                // if the queue overflows and no limit was specified, bail out
```
it's valid, since we throw an exception, so I'll just rephrase.

astefan replied:
You're right, I didn't look properly at `onFailure` method. Please, ignore.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java:
```diff
@@ -248,31 +205,21 @@ private void doResponse(RowSet rowSet) {
+                // if the queue overflows and no limit was specified, bail out
+                if (data.insertWithOverflow(new Tuple<>(row, counter.getAndIncrement())) != null && noLimit) {
+                    onFailure(new SqlIllegalArgumentException(
+                        "The default limit [{}] for aggregate sorting has been reached; please specify a LIMIT", data.size()));
```
it's the same actually, since we overflow then the size of the queue (initially set to MAXIMUM_SIZE) is reached. Will changed that though, for more clarity, thx!

matriv committed:
address comments

matriv commented:
@astefan Thanks for review, addressed comments.
Also there are tests where columns are presorted by ES: https://github.com/elastic/elasticsearch/blob/8124d3ab9add81b35cdc8a90341e9d05390d1434/x-pack/plugin/sql/qa/src/main/resources/agg-ordering.sql-spec#L80

matriv committed:
Change test name

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/execution/search/QuerierTests.java:
```diff
@@ -0,0 +1,117 @@
+        assertEquals(queueSize, results.size());
+        expected.sort((o1, o2) -> {
+            for (int j = 0; j < noColumns; j++) {
+                if (ordering[j]) { // asc
```
Can the content of the loop be changed to the following?

```
                    if (o1.get(j) < o2.get(j)) {
                        return ordering[j] ? -1 : 1;
                    } else if (o1.get(j) > o2.get(j)) {
                        return ordering[j] ? 1 : -1;
                    }
```

matriv replied:
Sure, thx! Applied it.

astefan reviewed (approved):
Great test. LGTM. Left one small comment.

matriv committed:
simplify test
